File: android/app/src/main/java/com/clipbridge/app/crypto/Crypto.kt
package com.clipbridge.app.crypto

import android.util.Base64
import com.iwebpp.crypto.TweetNaclFast
import java.security.SecureRandom

/**
 * Thin wrapper around TweetNaCl (X25519 key exchange + XSalsa20-Poly1305).
 * Mirrors the desktop crypto.ts implementation so both sides are compatible.
 */
object Crypto {

    data class KeyPair(val publicKey: String, val secretKey: String)

    fun generateKeyPair(): KeyPair {
        val kp = TweetNaclFast.Box.keyPair()
        return KeyPair(
            publicKey = Base64.encodeToString(kp.publicKey, Base64.NO_WRAP),
            secretKey = Base64.encodeToString(kp.secretKey, Base64.NO_WRAP),
        )
    }

    fun generateDeviceId(): String {
        val bytes = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val hex = bytes.joinToString("") { "%02x".format(it) }
        return "${hex.slice(0..7)}-${hex.slice(8..11)}-4${hex.slice(13..15)}-" +
               "${((hex[16].digitToInt(16) and 0x3) or 0x8).toString(16)}${hex.slice(17..19)}-${hex.slice(20..31)}"
    }

    data class EncryptedPayload(val nonce: String, val box: String)

    fun encrypt(plaintext: String, mySecretKey: String, peerPublicKey: String): EncryptedPayload {
        val box = TweetNaclFast.Box(
            Base64.decode(peerPublicKey, Base64.NO_WRAP),
            Base64.decode(mySecretKey,   Base64.NO_WRAP),
        )
        val nonce = box.generateNonce()
        val msg   = plaintext.toByteArray(Charsets.UTF_8)
        val encrypted = box.box(msg, nonce) ?: error("Encryption failed")
        return EncryptedPayload(
            nonce = Base64.encodeToString(nonce,     Base64.NO_WRAP),
            box   = Base64.encodeToString(encrypted, Base64.NO_WRAP),
        )
    }

    fun decrypt(payload: EncryptedPayload, mySecretKey: String, peerPublicKey: String): String? {
        return try {
            val box = TweetNaclFast.Box(
                Base64.decode(peerPublicKey, Base64.NO_WRAP),
                Base64.decode(mySecretKey,   Base64.NO_WRAP),
            )
            val decrypted = box.open(
                Base64.decode(payload.box,   Base64.NO_WRAP),
                Base64.decode(payload.nonce, Base64.NO_WRAP),
            ) ?: return null
            String(decrypted, Charsets.UTF_8)
        } catch (e: Exception) {
            null
        }
    }
}

File: desktop/src/lib/crypto.ts
/**
 * Crypto layer — X25519 key exchange + XSalsa20-Poly1305 (via TweetNaCl)
 * Everything here runs client-side only. Server never sees plaintext.
 *
 * NOTE: We use the native TextEncoder/TextDecoder instead of tweetnacl-util's
 * encodeUTF8/decodeUTF8 because the latter returns a plain String in some Vite/ESM
 * environments, which causes nacl.box.after to throw "unexpected type, use Uint8Array".
 */
import nacl from "tweetnacl";
import { encodeBase64, decodeBase64 } from "tweetnacl-util";

export interface KeyPair {
  publicKey: string;  // base64
  secretKey: string;  // base64
}

const encoder = new TextEncoder();
const decoder = new TextDecoder();

export function generateKeyPair(): KeyPair {
  const kp = nacl.box.keyPair();
  return {
    publicKey: encodeBase64(kp.publicKey),
    secretKey: encodeBase64(kp.secretKey),
  };
}

export function generateDeviceId(): string {
  const bytes = nacl.randomBytes(16);
  const hex = Array.from(bytes).map(b => b.toString(16).padStart(2, "0")).join("");
  return [
    hex.slice(0, 8),
    hex.slice(8, 12),
    "4" + hex.slice(13, 16),  // version 4
    ((parseInt(hex[16], 16) & 0x3) | 0x8).toString(16) + hex.slice(17, 20),
    hex.slice(20, 32),
  ].join("-");
}

/** Derive a shared key from our secret key + peer's public key */
function sharedKey(mySecretKey: string, peerPublicKey: string): Uint8Array {
  return nacl.box.before(decodeBase64(peerPublicKey), decodeBase64(mySecretKey));
}

export interface EncryptedPayload {
  nonce: string;   // base64
  box: string;     // base64
}

export function encrypt(
  plaintext: string,
  mySecretKey: string,
  peerPublicKey: string
): EncryptedPayload {
  const key = sharedKey(mySecretKey, peerPublicKey);
  const nonce = nacl.randomBytes(nacl.box.nonceLength);
  const msg = encoder.encode(plaintext);        // TextEncoder → always Uint8Array ✓
  const encrypted = nacl.box.after(msg, nonce, key);
  return {
    nonce: encodeBase64(nonce),
    box: encodeBase64(encrypted),
  };
}

export function decrypt(
  payload: EncryptedPayload,
  mySecretKey: string,
  peerPublicKey: string
): string | null {
  try {
    const key = sharedKey(mySecretKey, peerPublicKey);
    const decrypted = nacl.box.open.after(
      decodeBase64(payload.box),
      decodeBase64(payload.nonce),
      key
    );
    if (!decrypted) return null;
    return decoder.decode(decrypted);           // TextDecoder → always works ✓
  } catch {
    return null;
  }
}
